utils/shmem.c

```diff
@@ -0,0 +1,96 @@
+#include <errno.h>
+#include <fcntl.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "utils/shmem.h"
+
+#ifdef __ANDROID__
+
+int uftrace_shmem_open(const char *name, int oflag, mode_t mode)
+{
+	const char *uftrace_dir;
+	char *fname;
+	int fd;
+	int status;
+
+	uftrace_dir = uftrace_shmem_root();
+
+	status = mkdir(uftrace_dir, mode);
+	if (status < 0 && errno != EEXIST)
+		return -1;
+
+	if (asprintf(&fname, "%s/%s", uftrace_dir, name) < 0)
+		return -1;
+
+	fd = open(fname, oflag, mode);
+	if (fd >= 0) {
+		int flags = fcntl(fd, F_GETFD, 0);
+		flags |= FD_CLOEXEC;
+		if (fcntl(fd, F_SETFD, flags) < 0) {
+			int saved_errno = errno;
+			close(fd);
+			fd = -1;
+			errno = saved_errno;
+		}
```

Comment:
I think we can do this for non-Android as well.  But it should be a separate change.

Also I guess it can be a source of system-wide memory leak if it's not collected by uftrace after exec.  We need to make sure it doesn't leak the shared memory.